### ISSUE TITLE
UpToDate "set of project items was changed more recently" infinite loop (DLL stays the same, LastItemsChangedAtUtc stays the same)

### ISSUE THREAD

**Issue by tompazourek** [dotnet/project-system#6647]
**Visual Studio Version**: 16.7.3

**Summary**: I have a very large solution and some projects can somehow get into this weird state where they get built over and over even when they don't have any changes. For example, let's say `MyProject` is stuck in this state. I build it and get this log message:

```
FastUpToDate: The set of project items was changed more recently (2020-10-02 13:01:00) than the earliest output 'xxx\obj\Debug\MyProject.dll' (2020-10-02 13:00:58), not up to date. (MyProject)
```

Then Visual Studio builds the project, but the build doesn't actually end up doing anything, it only copies the already existing DLL that has the same `2020-10-02 13:00:58` timestamp. So next time I build it, the DLL timestamp is still the same, it'll still be older than the timestamp `2020-10-02 13:01:00` that's remembered by Visual Studio.

The log message comes from:

https://github.com/dotnet/project-system/blob/519340c5d37ab96aaddadb5eca93a5955831a869/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs#L242

And by reading the code, I just think that this process is faulty. The `LastItemsChangedAtUtc` never gets updated, the project gets built, but the DLL is only copied, so the DLL timestamp never gets updated, and I get stuck in a loop. I don't know exactly how it is that the `LastItemsChangedAtUtc` timestamp is 2 seconds after the timestamp of the DLL, but I guess it can happen in some way (maybe the items are loaded with some delay if the solution is big? maybe something caused some "reload"?). Anyway, there needs to be some circuit breaker to get out of this endless loop. Both timestamps stay the same forever, and the build just happens over and over. **Maybe a solution would be that the `LastItemsChangedAtUtc` is somehow reset when the project starts building?**

**Steps to Reproduce**: Unfortunately, I don't have a reliable way of reproducing this, it doesn't happen every time. On my big solution, it happens often, but for different projects each time, and I don't exactly know what determines that. But it's still a situation with those 2 timestamps that can happen, and there should be a way how to break out of the loop.

Some notes regarding reproducing this: I didn't modify any code files, only opened a solution, clicked Build, and then watched how some projects get stuck with those timestamps. I have automated restore of NuGet packages enabled (maybe it's somehow related to the strange `LastItemsChangedAtUtc` value).

**Expected Behavior**: If a project is not up-to-date, then gets built, it should be up-to-date from that point.

**Actual Behavior**: A project can get stuck not being up-to-date forever, which makes it (and all of it dependants) built forever.

**User Impact**: The builds are a lot slower on a very large solution (over 100 projects) if a project somewhere deep in the dependency tree gets stuck in this loop. This just destroys the development experience making developers wait a long time for every tiny change.

---

Could be related to #5794, #5968, #6129.

**Comment by tompazourek** [comment]
Update: Found a repro case. This is not exactly what happened on my solution (I didn't modify any files), but it's a way to get stuck into this infinite loop forever.

Here it is:

- Create a blank .NET Core Console App from the template
- Build it in Visual Studio until it stops building it and considers it up-to-date
- In Windows Explorer, copy and paste the `Program.cs` file into the same folder as `Program - Copy.cs`.
- Look inside Visual Studio, it should notice the new file `Program - Copy.cs` automatically.
- Undo in Windows Explorer (CTRL+Z) which makes the `Program - Copy.cs` disappear.
- Press build in Visual Studio, and voilà, the up-to-date check is broken because the `LastItemsChangedAtUtc` is newer, and the DLL just gets copied and doesn't get a new timestamp.

**Comment by drewnoakes** [comment]
I wonder if we should model the set of items last checked vs the last time they changed. This would allow us to produce a more helpful log message too, in the case where items changed (we could explain which were added/removed).

**Comment by tompazourek** [comment]
@drewnoakes That's not a bad idea, a system based on that approach might work better. However, for me, the problem isn't really that the system is imperfect in this regard. What I think is a bigger issue is that you can get stuck in this loop where the project keeps building over and over because the DLL is just copied with the same old timestamp. I think designing the algorithm to ensure that it never gets stuck in a loop is more important. If there are some circumstances where it builds when it shouldn't really need to, it's not great, but it's not as bad if it only does it once. **Having an algorithm that never gets stuck in a loop by design would feel like a much more solid solution.**

**Comment by drewnoakes** [comment]
Agreed, but I don't believe that's possible. Note that ultimately MSBuild decides whether to actually build or not, and it uses its own authoritative up-to-date check. In VS we don't have visibility into all the state that MSBuild uses to make its decision. Visual Studio's _fast_ up-to-date check (which we're talking about here) exists only because calling MSBuild is expensive, so we try to catch the bulk of cases where we believe nothing's changed and avoid calling MSBuild. In cases where we cannot be sure, we have to favor failing with extra builds over failing where no build occurs (which is much worse for the developer).

**Comment by tompazourek** [comment]
@drewnoakes I don't mind the extra builds if VS cannot be sure, what I see as a problem is if I get extra builds forever without any circuit breaker because none of the timestamps change.

I don't share your pessimism that getting out of the endless build loop is impossible, I think there has to be some way.

My suggestion is to use a timestamp of when was the last time VS triggered the MSBuild, e.g. `LastTimeBuildTriggeredUtc`. And instead of having:

```csharp
if (earliestOutputTime < state.LastItemsChangedAtUtc)
{
    return log.Fail("Outputs", "The set of project items was changed more recently ({0}) than the earliest output '{1}' ({2}), not up to date.", state.LastItemsChangedAtUtc, earliestOutputPath, earliestOutputTime);
}
```

We could have something like (just a rough idea):

```csharp
if (state.LastTimeBuildTriggeredUtc < state.LastItemsChangedAtUtc)
{
    return log.Fail("Outputs", "The set of project items was changed more recently ({0}) than the last build was triggered ({1}), not up to date.", state.LastItemsChangedAtUtc, state.LastTimeBuildTriggeredUtc);
}
```

This would make more sense. Because the `LastItemsChangedAtUtc` is a timestamp of VS action and we were comparing it to file modified timestamps. I see that as comparing apples to oranges. We should compare it to a timestamp like the `LastTimeBuildTriggeredUtc` (note: this doesn't exist yet, I made it up), then we'd be comparing apples to apples.

**Comment by jjmew** [comment]
@ocallesp We thinking getting additional logging will be helpful. Can you provide instructions to @tompazourek on how to do that?

**Comment by drewnoakes** [comment]
@tompazourek thanks for your analysis. The challenge we have is that we don't know when a build actually occurs, only that someone asked us whether things were up to date. In a previous version of the component, we tracked the last time we were asked, and considered that timestamp in much the same way you describe. However this failed in the case that we are asked if the project was up to date, but a build then did not occur. Once in that state, there's another infinite loop of appearing up to date.

Your repro steps (build, add item, remove item, build) make sense and we know how to fix that particular bug. We can and should improve the logging around changes to the project's items at the same time.

**Comment by tompazourek** [comment]
@drewnoakes Thanks for sharing more background information, and for trying to figure out some solution. There's still one thing that bugs me a bit. In my original problem, I don't do the "add item"/"remove item" thing. It just gets into the loop by itself without me modifying any files. And it's a still mystery to me as to why that happens. My theory is that it has either something to do with the automatic restore functionality in VS, or that the projects are loaded with some sort of delay because there's so many of them (very big solution). If you've got any ideas what to try to pin this down, this would be helpful as well, because it's the original problem that's actually bugging me. It appears to me that it happens sort of unexpectedly and randomly which is bad.

**Comment by drewnoakes** [comment]
> I don't know how yet, but maybe we could somehow detect that this copying happened and include the information in our timestamp comparisons.

I'm not sure how we'd do this either.

In the past I've wondered if we need a new handshake with MSBuild that provides the information we need so we don't have to try and infer it. I don't yet know how this might work.

> In my original problem, I don't do the "add item"/"remove item" thing. It just gets into the loop by itself without me modifying any files

I still think we could retain information about how the set of items changed to include in the output message. I suspect that would help identify what's triggering your original problem.

> What about abandoning the whole LastItemsChangedAtUtc timestamp altogether

Doing that would introduce a bug where:

- Build project
- Remove file from disk
- Build again
  - no build occurs as inputs all before outputs

**Comment by drewnoakes** [comment]
@tompazourek #6660 logs information about which items were added/removed/changed. Perhaps you could try it out on your solution. If you clone the branch and F5 it in the latest VS Preview then enable "info" logging, you should see details of what's changed. I say _should_ because I'm not 100% sure that this repo works at the moment with the public preview build of VS. Sometimes we take dependencies upon unreleased bits from other teams.

**Comment by tompazourek** [comment]
@drewnoakes Thanks for the info, the approach you suggested to start with extended logging makes sense. I'll have to find some time and set it up on my machine, and get back to you with more details.

---

> In the past I've wondered if we need a new handshake with MSBuild that provides the information we need so we don't have to try and infer it. I don't yet know how this might work.

Yes, I understand that this might actually need a modification in MSBuild to somehow share the information that some action has happened. Can you share some information about how the current "handshake with MSBuild" works? I know that there are also some `*.CopyComplete` files in the intermediate output, but I don't know the details.

--- 

Regarding abandoning the timestamp `LastItemsChangedAtUtc`, maybe we could change it to `LastItemsRemovedAtUtc` and it would have less issues. Any thoughts on that? I know it's not that simple, I'm just thinking about ideas...

**Comment by drewnoakes** [comment]
@tompazourek did you get a chance to try out a build from that PR?

**Comment by tompazourek** [comment]
@drewnoakes I just tried now, it but failed.

- Installed VS Preview.
- Cloned [your fork](https://github.com/drewnoakes/project-system/).
- Checked out [your branch](https://github.com/drewnoakes/project-system/tree/up-to-date-check-item-change-logging).
- Opened the solution in VS Preview.
- Installed additional components into VS Preview that it asked me to (I think one of them was the VS SDK)
- F5 doesn't do anything, there doesn't appear to be any project that could be run, it's all just class libraries.
- The solution doesn't even build on my machine yet. The output just says `Build started...` and it looks to me like it's stuck on package restore as the status bar just says **Retrying 'FindPackagesByIdAsync'**.

I don't have too much time lately, unfortunately, I'm quite busy with my job now. But I hope I find more time soon to figure out how to set it up.

Did you try this yourself, is it working for you?

**Comment by drewnoakes** [comment]
The startup project must be `ProjectSystemSetup`. Building that project causes a VSIX to be deployed to your local experimental hive. When you F5 or Ctrl+F5 that project, a new VS will launch that includes that version of the .NET Project System. Your regular VS install is left unmodified.

Yes I tried this and it works for me. The reason to run it on your machine is to diagnose your original problem, to learn what item(s) are being added/removed from your projects over time, resulting in the up-to-date check requesting more builds than you suspect are necessary. Once we know which items are involved in that, we can determine whether there is a product bug to address.

Note too that you'll need to set the up-to-date log level to "info" or higher.

**Comment by tompazourek** [comment]
@drewnoakes Thanks for explaining. I'll try it again and get back to you with more information.

**Comment by tompazourek** [comment]
I managed to get this working now and tried it. I must say that the logging is great. I'd like it very much if your PR would actually get merged and became a part of VS.

But I have bad news for this issue. I wasn't able to reproduce getting into the loop on my large solution in the VS preview. I've then tried to reproduce the issue again in my stable Visual Studio version (I currently use the 16.7.6 instead of the original 16.7.3), and I still experience the issue.

- Maybe it's some other change between the 16.7.6 and the 16.8 preview 3.2 that improved this.
- Maybe it's some extensions that I have in my VS that I don't have on the preview (like ReSharper). **I can still test this more, and get back to you with more information, but it will take some time.**

Is it possible to eventually get the improved logging PR into the stable VS? I'll continue monitoring this issue and provide more information.

**Comment by drewnoakes** [comment]
Thanks for following up. This logging will be useful so we'll get it merged in one form or another, but not likely in 16.8 unfortunately.

If you're motivated, you could try the same steps in the 16.7 version of VS. If it fails due to mismatched APIs, you could rebase my PR's branch on the [dev16.7.x](https://github.com/dotnet/project-system/tree/dev16.7.x) branch in order to identify what changes are causing the problem you're seeing.

**Comment by tompazourek** [comment]
@drewnoakes Thanks, that's a good idea. I think I'll try that once I find some free time again, it's interesting stuff :)

**Comment by drewnoakes** [comment]
@tompazourek thanks. I don't see any way for us to progress this issue without the log information from your local repro. Let us know when you have more info.

**Comment by tompazourek** [comment]
@drewnoakes I just tried it, cherry-picked the changes on top of `dev16.7.x` and tried it there. And I was able to consistently reproduce it again there. So it seems to be only an issue in 16.7, and not 16.8, which is great. I'd still like to know what it was though to be sure.

However, I don't get any of the extra logs in the output, I only get the `FastUpToDate: The set of project items was changed more recently` message, but I don't get the specific items listed after that (it appears that the `state.LastItemChanges` is empty). I'll try to trace this further and update if I figure out something more.
